src/DurableTask.Netherite.AzureFunctions/NetheriteProviderFactory.cs

```diff
@@ -48,6 +51,7 @@ public class NetheriteProviderFactory : IDurabilityProviderFactory
         public NetheriteProviderFactory(
             IOptions<DurableTaskOptions> extensionOptions,
             ILoggerFactory loggerFactory,
+            IServiceProvider serviceProvider,
```

Comment:
Yep, constructor overloading sounds like a good way to avoid a breaking change.

Reply:
Is this really a breaking change? Adding an optional argument should not break any existing uses of the constructor.

Reply:
I agree though that it does seem like a good idea to not force users into using an IServiceProvider which could be inconvenient if they are not already using DI.

Reply:
Yes, it is a runtime breaking change - existing apps that drop in the version of this assembly will encounter `MethodMissingException` as this is a different ctor signature at runtime.

Reply:
This constructor is never called by any client code, it is only accessible via DI.
So this is not actually a breaking change.

Reply:
I was careful to avoid a breaking change for AF users.

I was not careful for DF users though, so I should add some overloads to keep things compatible.

Reply:
Many breaking changes can be runtime only. This will break if I take an app compiled against the previous version of DurableTask.Netherite and drop in the new dll. When following semver, you should always be able to drop in a minor version rev _without needing to recompile your application._

Reply:
Ok, got it. It means I cannot change method signatures.

Reply:
I think DI is based on runtime reflection, so I think it should work. I can test that.

Reply:
> When following semver, you should always be able to drop in a minor version rev without needing to recompile your application.

Supporting update without recompilation for _any_ possible situation seems like an unnecessarily high bar. I can see this being important for some projects, but it seems like overkill in this context if applied as an abstract rule. IMO it is enough to consider the supported ways in which our APIs may be used, and make sure to support updates for those supported scenarios. For example, we can assume that Netherite.AzureFunctions.dll is only used from Microsoft.Azure.WebJobs.Extensions.DurableTask.dll.

Also, I cannot imagine that there are many users that need to update Netherite without recompiling. That seems like a very fringe scenario. From what I have seen, any function app deployments are being recompiled right before being deployed.

Reply:
I have manually tested on the sample function app that I can drop in the new dlls (DurableTask.Netherite.AzureFunctions.dll and DurableTask.Netherite.dll) and the app still runs, without recompilation.

Reply:
> I also recommend to split these changes into a different constructor. This will avoid breaking changes and also if there is an either/or option (provide creds this way or a different way) the two ctors can represent the two different ways.

how does DI know which constructor to use if there is more than one?

Reply:
> how does DI know which constructor to use if there is more than one?

I see you have answered this below. There is an attribute for it.